Pranav

pranavaher-kylas opened the pull request.

Commit:
initial commit

Commit:
removed unnecessary files

pranavaher-kylas committed:
Added cart and spec file

pranavaher-kylas committed:
Added cart and spec files

Makwana-Jigar commented on pranav/cart/spec.rb:
```diff
@@ -0,0 +1,134 @@
+      expect_any_instance_of(Kernel).to receive(:gets).and_return("apple")
+      expect_any_instance_of(Kernel).to receive(:gets).and_return(60)
+      expect_any_instance_of(Kernel).to receive(:gets).and_return(15)
+      # add_to_menu("apple", menu, 60, 15)
```
you can use single statement for mocking gets like here it can be expect_any_instance_of(Kernel).to receive(:gets).and_return("apple\n", "60\n", "15\n")

pranavaher-kylas committed:
updated expect

MayurJain-Kylas commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+class Cart
+    def initialize
+      @menu = { 1 => { name: 'aa', price: 10, quantity: 10 }, 2 => { name: 'bb', price: 30, quantity: 10 }, 3 => { name: 'cc', price: 40, quantity: 7 }, 4 => { name: 'dd', price: 70, quantity: 10 }, 5 => { name: 'ee', price: 90, quantity: 10 } }
```
Rather than giving index as keys wouldnt it be more convinient to give item name as a key

pranavaher-kylas replied:
since in realtime databases object is being identifies by index, I have implemented it here.

MayurJain-Kylas replied:
Right and ID is mostly the unique attribute which usually is name and not index

MayurJain-Kylas commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      puts "Your total amount is: #{total_amount}"
+    end
+  
+    def add_to_menu(item_name, item_price, item_quantity)
```
Methods to modify menu should not be in Cart class

shubham-selldo replied:
Right

MayurJain-Kylas commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+    end
+  
+    def remove_item_from_cart(item_number)
+      if !@cart.key?(item_number)
```
Item name rather than number would be more functional

pranavaher-kylas replied:
As i have used hash for storing cart/menu and number being key and name, quantities, price as value, it makes hard to find object in cart and menu by names. Also for user it is easy to type item number instead of item name.

MayurJain-Kylas commented on pranav/cart/spec.rb:
```diff
@@ -0,0 +1,129 @@
+    it "item does not exists, in menu key for that item id should return false"       
+      expect_any_instance_of(Kernel).to receive(:gets).and_return(6)
+      expect_any_instance_of(Kernel).to receive(:gets).and_return(8)
+      add_to_cart(6,8, menu, cart)
```
Space after comma

MayurJain-Kylas commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+        return (amount * 0.2)
+      elsif amount > 100
+        return (amount * 0.1)
+      end
```
No space needed u can use else

pranavaher-kylas replied:
Changes updated.

MayurJain-Kylas commented on pranav/cart/spec.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+  end  
+end 
+
```
No space needed

pranavaher-kylas committed:
given space after comma and removed extra lines at the end of file.

pranavaher-kylas committed:
Update changes 

1. Add 2 space indentation
2. Use require against load
3. Add 1 blank line at the end of file
4. Give space before and after operators and space after comma

pranavaher-kylas committed:
Remove extra line and add else block

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+class Cart
+    def initialize
```
Use 2 space indentation everywhere.
Follow best practices.

pranavaher-kylas replied:
Done.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+class Cart
+    def initialize
+      @menu = { 1 => { name: 'aa', price: 10, quantity: 10 }, 2 => { name: 'bb', price: 30, quantity: 10 }, 3 => { name: 'cc', price: 40, quantity: 7 }, 4 => { name: 'dd', price: 70, quantity: 10 }, 5 => { name: 'ee', price: 90, quantity: 10 } }
```
Try to use real time names like 'Apple', 'Orange' instead of 'aa', 'bb'.

pranavaher-kylas replied:
Done

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+class Cart
+    def initialize
+      @menu = { 1 => { name: 'aa', price: 10, quantity: 10 }, 2 => { name: 'bb', price: 30, quantity: 10 }, 3 => { name: 'cc', price: 40, quantity: 7 }, 4 => { name: 'dd', price: 70, quantity: 10 }, 5 => { name: 'ee', price: 90, quantity: 10 } }
```
try to be consistent.
like if you are using => in hashes.. use it everywhere.
we prefer `:`

pranavaher-kylas replied:
used => everywhere now

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      @cart_counter = @cart.size
+    end
+  
+    def cal_gst(amount)
```
use proper names like `calculate_gst`

pranavaher-kylas replied:
Done.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+  
+    def cal_gst(amount)
+      gst_amount = 0.0
+      s_gst = 8
```
s_gst and c_gst are constant values.
so, move these to constants

pranavaher-kylas replied:
Done.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      gst_amount
+    end
+  
+    def cal_discount(amount)
```
same here `calculate_discount`

pranavaher-kylas replied:
Done.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+  
+    def cal_discount(amount)
+      puts 'We have 10%, 20% and 30% discount based on your bill'
+      if amount > 300
```
this can be refactored.
avoid returning multiple times.

pranavaher-kylas replied:
Removed multiple return statements.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+    def cal_gross_bill
+      gross_amount = 0.0
+      @cart.each do |_key, value|
+        gross_amount += (value[:price] * value[:quantity]).to_i
```
any reason for using `to_i` here ?

pranavaher-kylas replied:
Fixed it.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      gross_amount
+    end
+  
+    def make_bill
```
All this billing related actions should not be part of `Cart` class.
Ideally Cart should only have actions like add to cart, remove from cart.

And billing, menu should be separated. Please make necessary changes

pranavaher-kylas replied:
Created separate class for cart and menu.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      puts
+    end
+  
+    def print_store(store)
```
this is also irrelevant to Cart class.

pranavaher-kylas replied:
Fixed.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+    end
+  
+    def add_to_cart(item_number, item_quantity)
+      if !@menu.key?(item_number)
```
we can use here `unless` to avoid negative conditions

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+  
+    def add_to_cart(item_number, item_quantity)
+      if !@menu.key?(item_number)
+        puts 'We dont have any such item'
```
this statement can be more informative and grammatically correct.

pranavaher-kylas replied:
Fixed.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      if !@menu.key?(item_number)
+        puts 'We dont have any such item'
+        nil
+      elsif item_quantity <= @menu[item_number][:quantity]
```
instead of writing large chunk of logic in if elsif conditionals, we can make use of return statements.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+    def add_to_cart(item_number, item_quantity)
+      if !@menu.key?(item_number)
+        puts 'We dont have any such item'
+        nil
```
why nil here ?

pranavaher-kylas replied:
Fixed

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      else
+        puts "#{@cart[item_number][:name]} removed from cart: "
+        @cart.delete(item_number)
+        @cart_counter -= 1
```
shouldn't we add respective quantity in menu as well ?

pranavaher-kylas replied:
Done

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+    def remove_item_from_cart(item_number)
+      if !@cart.key?(item_number)
+        puts 'You dont have any such item'
+        nil
```
same comments as for add_to_cart

pranavaher-kylas replied:
Done.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      puts '--------Welcome to shopping cart--------'
+      puts
+      puts 'Press -1 to exit'
+      puts '1. Add item in menu'
```
this can definitely be refactored. please check.

pranavaher-kylas replied:
Refactored. Removed nested loops.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+      puts
+    end
+  
+    def driver
```
This is not expected in Cart class.
Do not hesitate to use multiple classes.
Refactor all actions accordingly.

pranavaher-kylas replied:
Created separate classes.

shubham-selldo commented on pranav/cart/main.rb:
```diff
@@ -0,0 +1,192 @@
+          loop do
+            print 'Enter item name to add to menu: '
+            item_name = gets.chomp
+  
```
avoid unnecessary empty lines

pranavaher-kylas replied:
Fixed.

shubham-selldo commented on pranav/cart/spec.rb:
```diff
@@ -0,0 +1,123 @@
+        c_gst = 8
+      end      
+      it "return GST amount on gross_amount" 
+        true_amount = (gross_amount / 100) * (c_gst + s_gst)
```
in test cases, we should not use dynamic values.
we have to use static values to compare against actual values.

pranavaher-kylas replied:
Fixed.

shubham-selldo commented on pranav/cart/spec.rb:
```diff
@@ -0,0 +1,123 @@
+      end
+    end  
+  end  
+end 
```
have you executed your test cases?
please attach output of the same.

pranavaher-kylas replied:
Done.

![Screenshot from 2023-03-05 15-37-31](https://user-images.githubusercontent.com/125980413/222955103-e2ffee63-6012-4931-9c13-cbb9c242f565.png)

shubham-selldo reviewed (changes requested):
Test cases have major syntax error.
Please attach screen shot of rspecs

pranavaher-kylas committed:
Update main.rb

Separate logic for Cart and Menu.
Add space and line indention. 
Use proper names.
Refactor code.

pranavaher-kylas committed:
Update spec.rb

pranavaher-kylas committed:
line indention fixed

pranavaher-kylas committed:
Fix variable names
